packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts

```diff
@@ -327,12 +364,121 @@ export class AutomergeHost extends Resource {
+  }
+
+  refreshCollection(collectionId: string) {
+    this._collectionSynchronizer.refreshCollection(collectionId);
+  }
+
+  async getCollectionSyncState(collectionId: string): Promise<CollectionSyncState> {
+    const result: CollectionSyncState = {
+      peers: [],
+    };
+
+    const localState = this.getLocalCollectionState(collectionId);
+    const remoteState = this.getRemoteCollectionStates(collectionId);
+
+    if (!localState) {
+      return result;
+    }
+
+    for (const [peerId, state] of remoteState) {
+      const diff = diffCollectionState(localState, state);
+      result.peers.push({
+        peerId,
+        differentDocuments: diff.different.length,
+      });
+    }
+
+    return result;
+  }
+
+  /**
+   * Update the local collection state based on the locally stored document heads.
+   */
+  async updateLocalCollectionState(collectionId: string, documentIds: DocumentId[]) {
+    const heads: Record<DocumentId, Heads> = {};
+    await Promise.all(
+      documentIds.map(async (documentId) => {
+        heads[documentId] = (await this.getHeads(documentId)) ?? [];
+      }),
+    );
+    this._collectionSynchronizer.setLocalCollectionState(collectionId, { documents: heads });
+  }
+
+  private _onCollectionStateQueried(collectionId: string, peerId: PeerId) {
+    this._collectionSynchronizer.onCollectionStateQueried(collectionId, peerId);
+  }
+
+  private _onCollectionStateReceived(collectionId: string, peerId: PeerId, state: unknown) {
+    this._collectionSynchronizer.onRemoteStateReceived(collectionId, peerId, decodeCollectionState(state));
+  }
+
+  private _queryCollectionState(collectionId: string, peerId: PeerId) {
+    this._echoNetworkAdapter.queryCollectionState(collectionId, peerId);
+  }
+
+  private _sendCollectionState(collectionId: string, peerId: PeerId, state: CollectionState) {
+    this._echoNetworkAdapter.sendCollectionState(collectionId, peerId, encodeCollectionState(state));
+  }
+
+  private _onPeerConnected(peerId: PeerId) {
+    this._collectionSynchronizer.onConnectionOpen(peerId);
+  }
+
+  private _onPeerDisconnected(peerId: PeerId) {
+    this._collectionSynchronizer.onConnectionClosed(peerId);
+  }
+
+  private _onRemoteCollectionStateUpdated(collectionId: string, peerId: PeerId) {
+    const localState = this._collectionSynchronizer.getLocalCollectionState(collectionId);
+    const remoteState = this._collectionSynchronizer.getRemoteCollectionStates(collectionId).get(peerId);
+
+    if (!localState || !remoteState) {
+      return;
+    }
+
+    const { different } = diffCollectionState(localState, remoteState);
+
+    if (different.length === 0) {
+      return;
+    }
+
+    log.info('replication documents after collection sync', {
+      count: different.length,
+    });
+
+    // Load the documents that are different.
+    for (const documentId of different) {
+      this._repo.find(documentId);
+    }
+  }
+
+  private _onHeadsChanged(documentId: DocumentId, heads: Heads) {
```

Comment:
should it maybe happen afterSave? we don't use batch in any way for atomicity here, but if save fails we have non-existent heads recorded